### PR TITLE
feat(attachments): rework edit attachments ui

### PR DESCRIPTION
**Pull request by lepokle**
In order to switch to jquery datatable and to have a common way to insert data into that table
(on initialization and on adding attachments) even the initial data is now loaded via ajax. The
used format is JOSN.

## Important Notes:
- jquery DataTables has been upgraded to version 1.10.15. The feature 'rowId' was needed to be able to define a custom value to be the row id
- the AttachmentAwarePortlet now sends pure JSON for the following requests:
    - AttachmentList
    - AttachmentLinkTo

## The attachments feature is currently used for the following main objects:
- projects
- components
- releases

The edit attachment feature consists of two parts: editing metadata of attachments as well as uploading new files.
Uploading of files is - as a feature - independent from editing metada. This is also respected by the code now.

## Code reorganisation:
- The module for uploading attachments is now {html,js}/utils/includes/attachmentsUpload.{jspf,js}.
  The corresponding javascript code will now call a callback with a json representation of every uploaded attachment. No DOM manipulations outside of its own dialog anymore.
- The module for editing attachments is now frontend/sw360-portlet/src/main/webapp/html/utils/includes/editAttachments.jspf only. No other includes necessary.
  However there are some preconditions that must be fulfilled: 1) requireJS must be laoded 2) Some classes must be imported, see descriptive headrer of the file
- The following files are now obsolete and have been deleted. Necessary code have been copied to one of the files above:
    - html/utils/ajax/attachmentsAjax.jsp (not needed anymore because the server sends pure JSON now)
    - html/utils/includes/attachmentsDelete.jsp (integrated into editAttachments.jspf, was not able to live alone)
    - html/utils/includes/attachmentsUpload.jsp (moved to html/utils/includes/attachmentsUpload.jspf and modularized)
    - frontend/sw360-portlet/src/main/webapp/js/resumableAttachments.js (integrated into js/utils/includes/attachmentsUpload.js, was not able to live alone)

## Styling changes:
- input and select fields in DataTables are no longer wider than their column
- there is a new class "added" which will set the background to green for a row in a DataTable
- there will be vertical lines in the table header, if a header cell spans more than one row
- vertical alignment in the attachment info table is now middle

## New modules:
The module "modules/datatable-renderers" has been added which contains some useful renderes to be used with a datatable. The following renderes have been added:
- ellipsis: renders a text with an ellipsis if necessary
- inputSelect: renders a select field
- inputText: renders a text field
- actions: renders actions
Please the the file js/modules/datatable-renderers.js for more information.

## Other changes:
- remove unused constant in PortalConstants

### CODE REVIEW

File: frontend/sw360-portlet/src/main/webapp/js/modules/datatables-renderer.js [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+     *		  	}
+     *		}
+     *  </code>
+     */
+    $.fn.dataTable.render.inputText.useInputDialog = function(td, dialogTitle) {
+        var $text = $(td).find('input');
+        $text.on('focus', function() {
+            $text.blur(); // unfocus otherwise the dialog will immediately open again after closing
+
+            $('<div><input type="text" style="margin-top: 10px; width: 90%;"/></div>').dialog({
```
Comment (lepokle) [inline frontend/sw360-portlet/src/main/webapp/js/modules/datatables-renderer.js]: As far as I know you can't use input fields in jquery-confirm dialogs. At least I didn't find any hint in the documentation.

Comment (alexbrdn) [inline frontend/sw360-portlet/src/main/webapp/js/modules/datatables-renderer.js]: Yes, you can. Here's their example:
https://craftpip.github.io/jquery-confirm/
image

Comment (lepokle) [inline frontend/sw360-portlet/src/main/webapp/js/modules/datatables-renderer.js]: Thank you, I was on the documentation for jquery.confirm not on jquery-confirm. Grr.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/AttachmentAwarePortlet.java [hunk 2]
```diff
@@ -95,27 +109,63 @@ protected void dealWithAttachments(ResourceRequest request, ResourceResponse res
...
+        }).map(attachment -> {
+            try {
+                return serializer.toString(attachment);
+            } catch (TException exception) {
+                return "{ \"error\": \"" + TagUtils.escapeAttributeValue(exception.toString()) + "\" }";
+            }
+        }).collect(Collectors.toList());
+
+        // enum data to be able to interpret attachment type
+        JSONObject attachmentTypes = JSONFactoryUtil.createJSONObject();
```
Comment (lepokle) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/AttachmentAwarePortlet.java]: Is it possible that Jackson is not a dependency of the portlet project yet? In this case I need to add it. Is this ok?

Comment (lepokle) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/AttachmentAwarePortlet.java]: Ah, ok, comes with ektorp. I'll see what is possible.

**Comment by lepokle** [comment]
@alexbrdn Please check again :-)

**Comment by mcjaeger** [comment]
(from testing in group)
pls. let's display the comment field disabled unless the user selects a state (approved or rejected). Jumping back to the default state (not approved from approved or rejected) shall remove the comment (in the backend) and the user e-mail and the modification date.

**Comment by lepokle** [comment]
Added commented functionality from @mcjaeger 
In addition the text that will be deleted will be displayed in red.

**Comment by mcjaeger** [comment]
testing was not OK:
1. for projects and releases I was not able to add attachments because clicking the button "Update ..." does not work  ("Cancel" does though).
2. It seems to be connected with the change of the approval status
3. When clicking the Update Release button a strange alert message was shown in the browser (see screenshot)

propose to have a debug session. Maybe it is connected with the conflicting files (a missing update?)

036 oss sw360 20170823 screenshot fail attachments for releases

**Comment by lepokle** [comment]
Branch updated.
@mcjaeger cannot reproduce your issue, sorry.

**Comment by mcjaeger** [comment]
I think the issue was solved as part of #544 (and only seen on firefox), @lepokle could you pls. rebase

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java [hunk 3]
```diff
@@ -177,66 +179,97 @@ else if (field == User._Fields.USER_GROUP)
...
-                    //the sha1 checksum is not computed here, but in the backend, when updating the component in the database
-                    attachment = CommonUtils.getNewAttachment(user, id, fileNames[i]);
-                    documentAttachments.add(attachment);
-                }
-                attachment.setCreatedComment(createdComments[i]);
-                attachment.setAttachmentType(getAttachmentTypefromString(atypes[i]));
-                if(!checkedComments[i].equals(attachment.checkedComment) || attachment.checkStatus != getCheckStatusfromString(checkStatuses[i])){
+
+        if (ids == null || ids.length == 0) {
+            LOGGER.info("No ids transmitted. No attachments will be saved or deleted.");
```
[the remark this answers is not in the document]
Comment (lepokle) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: fixed, Thx!

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java [hunk 4]
```diff
@@ -177,66 +179,97 @@ else if (field == User._Fields.USER_GROUP)
...
+        }
+
+        Map<String, Attachment> documentAttachmentMap = documentAttachments != null
+                ? documentAttachments.stream().collect(Collectors.toMap(Attachment::getAttachmentContentId, Function.identity()))
+                : Maps.newHashMap();
+        for (int i = 0; i < ids.length; ++i) {
+            String id = ids[i];
+            Attachment attachment;
+
+            attachment = documentAttachmentMap.get(id);
```
[the remark this answers is not in the document]
Comment (lepokle) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: It's a map and I have no such method found for maps.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java [hunk 5]
```diff
@@ -177,66 +179,97 @@ else if (field == User._Fields.USER_GROUP)
...
+            return attachments;
+        } else if (CommonUtils.oneIsNull(fileNames, types, createdComments, checkStatuses, checkedComments)) {
+            LOGGER.error("Invalid request content. One of the attachment parameters is null.");
+            return documentAttachments;
+        } else if (!CommonUtils.allHaveSameLength(ids, fileNames, types, createdComments, checkStatuses, checkedComments)) {
+            LOGGER.error("Not all of the attachment parameter arrays have the same length! No attachments will be saved or deleted.");
+            return documentAttachments;
+        }
+
+        Map<String, Attachment> documentAttachmentMap = documentAttachments != null
```
[the remark this answers is not in the document]
Comment (lepokle) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: It's a map and I have no such method found for maps.

**Comment by lepokle** [comment]
Updated, @ChristophNiehoff  will retest the moderation stuff since this was the cause of most conflicts.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexbrdn) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: this one should do: https://github.com/sw360/sw360portal/blob/master/libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/CommonUtils.java#L151

[the remark this answers is not in the document]
Comment (lepokle) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java]: fixed, thx.

**Comment by ChristophNiehoff** [comment]
Tested the moderation stuff for functionality. It works.

**Pull request merged**